apps/gateway/lib/gateway/controllers/user_controller.ex

```diff
@@ -14,4 +15,22 @@ defmodule Gateway.Controllers.UserController do
       {:error, _changeset} -> send_resp(conn, 400, Jason.encode!(%{"error" => "failed to update"}))
     end
   end
+
+  def claim_daily_reward(conn, %{"user_id" => user_id}) do
+    with {:ok, user} <- Users.get_user(user_id),
+         {:ok, :can_claim} <- Rewards.user_claimed_today?(user),
+         {:ok, daily_reward} <- Rewards.claim_daily_reward(user),
+         {:ok, user} <-
+           Users.update_user(user, %{
+             last_daily_reward_claim: DateTime.utc_now(),
+             last_daily_reward_claim_type: daily_reward
+           }) do
+      send_resp(conn, 200, Jason.encode!(user.id))
+    else
+      {:error, :not_found} -> send_resp(conn, 404, Jason.encode!(%{"error" => "not found"}))
+      {:error, :already_claimed} -> send_resp(conn, 400, Jason.encode!(%{"error" => "already claimed"}))
+      {:error, :invalid_reward} -> send_resp(conn, 400, Jason.encode!(%{"error" => "invalid reward"}))
+      {:error, _changeset} -> send_resp(conn, 400, Jason.encode!(%{"error" => "failed to update"}))
+    end
```

Comment:
Let's try and improve this

First of ["Anti-pattern: complex else in with"](https://hexdocs.pm/elixir/code-anti-patterns.html#complex-else-clauses-in-with) for a brief explanation on why doing `with` like this is not a good idea for maintainability

Second ["Phoenix: Action fallback"](https://hexdocs.pm/phoenix/json_and_apis.html#action-fallback) explaining a more generic approach to dealing with this kind of things. Also good to read the following section ["FallbackController and ChangesetJSON"](https://hexdocs.pm/phoenix/json_and_apis.html#fallbackcontroller-and-changesetjson)

This means that ideally this function should end up more like
```elixir
  def claim_daily_reward(conn, %{"user_id" => user_id}) do
    with {:ok, user} <- Users.get_user(user_id),
         {:ok, :can_claim} <- Rewards.user_claimed_today?(user),
         {:ok, daily_reward} <- Rewards.claim_daily_reward(user),
         {:ok, user} <- Users.update_user(user, %{last_daily_reward_claim: DateTime.utc_now(), last_daily_reward_claim_type: daily_reward}) do
      send_resp(conn, 200, Jason.encode!(user.id))
    end
  end
```

Reply:
You're totally right about the with-else and I didn't know about the other two articles. I'll implement the 1st then read the others. Thanks a lot!